src/cyclebane/node_values.py

```diff
@@ -0,0 +1,302 @@
+    simple Python iterables.
+    """
+
+    @staticmethod
+    def from_array_like(values: Any, *, axis_zero: int = 0) -> ValueArray:
+        if hasattr(values, 'dims'):
+            return DataArrayAdapter(values)
+        if values.__class__.__name__ == 'ndarray':
+            return NumpyArrayAdapter(values, axis_zero=axis_zero)
+        return SequenceAdapter(values, axis_zero=axis_zero)
+
+    @abstractmethod
+    def sel(self, key: tuple[tuple[IndexName, IndexValue], ...]) -> Any:
+        """Return data by selecting from index with given name and index value."""
+
+    @abstractmethod
+    def __getitem__(self, key: int | slice | tuple[int | slice, ...]) -> ValueArray:
+        pass
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        pass
+
+    @property
+    @abstractmethod
+    def index_names(self) -> tuple[IndexName, ...]:
+        pass
+
+    @property
+    @abstractmethod
+    def indices(self) -> dict[IndexName, Iterable[IndexValue]]:
+        pass
+
+
+class PandasSeriesAdapter(ValueArray):
+    def __init__(self, series: 'pandas.Series', *, axis_zero: int = 0):
+        self._series = series
+        self._axis_zero = axis_zero
+
+    def sel(self, key: tuple[tuple[IndexName, IndexValue], ...]) -> Any:
+        if len(key) != 1:
+            raise ValueError('PandasSeriesAdapter only supports single index')
+        _, i = key[0]
+        return self._series.loc[i]
+
+    def __getitem__(
+        self, key: int | slice | tuple[int | slice, ...]
+    ) -> PandasSeriesAdapter:
+        return PandasSeriesAdapter(self._series[key], axis_zero=self._axis_zero)
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        return (len(self._series),)
+
+    @property
+    def index_names(self) -> tuple[IndexName, ...]:
+        return tuple(self.indices)
+
+    @property
+    def indices(self) -> dict[IndexName, Iterable[IndexValue]]:
+        name = (
+            self._series.index.name
+            if self._series.index.name is not None
+            else f'dim_{self._axis_zero}'
+        )
+        return {name: self._series.index}
+
+
+class DataArrayAdapter(ValueArray):
+    def __init__(
+        self,
+        data_array: 'xarray.DataArray | scipp.Variable | scipp.DataArray',
+    ):
+        self._data_array = data_array
+
+    def sel(self, key: tuple[tuple[IndexName, IndexValue], ...]) -> Any:
+        # Note: Eventually we will want to distinguish between dims without coords,
+        # where we will use a range index, and dims with coords, where we will use the
+        # coord as an index. For now everything is a range index.
+        # We use isel instead of sel, since we default to range indices for now.
+        if hasattr(self._data_array, 'isel'):
+            return self._data_array.isel(dict(key))
+        values = self._data_array
+        for label, i in key:
+            # This is Scipp notation, Xarray uses the 'isel' method.
+            values = values[(label, i)]
```

Comment:
```suggestion
            values = values[label, i]
```
?

Reply:
I used an explicit (rather than implicit) tuple since a reader of Cyclebane might not be familiar with Scipp. I though the notation without only a comma might be more confusing?

Reply:
I don't think this code is comprehensible to people who are unfamiliar with Scipp in any case. But ok.

Reply:
Adding a comment.